### ISSUE TITLE
Other Add-ons may accidentally duplicate template if they set from identity 

### ISSUE THREAD

**Issue by RealRaven2000** [RealRaven2000/SmartTemplate4#108]
I am just fixing a problem that is connected to setting the identity based on the folder (QuickFolders, see [this issue](https://github.com/RealRaven2000/QuickFolders/issues/91) for reference). This may / will lead to 2 different (or the same) template being displayed - essentially an account-specific template gets injected twice, without removing the first one.

I have seen template duplication before, which also can have been a problem created by other Add-ons such as Identity-picker or "right identity" - as these trigger an additional round of inserting the template. In my investigation I have found that it is usually caused by an unexpected separate call to Thunderbird's function LoadIdentity (which is overloaded / wrapped in SmartTemplates). In any case inserting the template should be more robust and detect against duplication like this.

**Comment by RealRaven2000** [comment]
This version should fix the issue:

[smartTemplate-fx-3.3pre28.zip](https://github.com/RealRaven2000/SmartTemplate4/files/5759139/smartTemplate-fx-3.3pre28.zip)

As always, to install download zip file and drag into Thunderbird Add-ons Manager / extensions pane.

**Comment by macdeport** [comment]
Really sorry to say that this fix don't solve this issue: OK/not OK depending of the precise mail, but for the moment it is unpredictable... at least by me.

**Comment by RealRaven2000** [comment]
> Really sorry to say that this fix don't solve this issue: OK/not OK depending of the precise mail, but for the moment it is unpredictable... at least by me.

Did you try to restart Thunderbird just to be 100% sure? It seems to work for me. If you can reproduce the problem with a special case I would like to know...

**Comment by macdeport** [comment]
> Did you try to restart Thunderbird just to be 100% sure?
Yes, for sure ;-)

Here is my **forward** smarttemplate in case:

![floating-screenshot-205816](https://user-images.githubusercontent.com/26068595/103445450-7521d180-4c74-11eb-8cbd-85d67a644b2e.jpg)

```
%header.set(priority,high)%%cursor%
--
%sig%

On %X:=sent%%d%/%m%/%Y% %H%:%M%:%S% %from(name)% <%from(mail)%> wrote:
> To: %to(name)% <%to(mail)%>[[
> Cc: %cc(name)% <%cc(mail)%>]]
> Subject: %subject%
> Message-ID: %Message-ID%
[[> Reply-to: %Reply-to%
]]>
```
I will send you a private mail with more stuff...

**Comment by RealRaven2000** [comment]
looks like you forward in plain text mode? i am using  a lot of html for transporting SmartTemplate's data. not sure whether the `body `element survives that...

**Comment by RealRaven2000** [comment]
by the way, you should put this part into the quote header field:

```
On %X:=sent%%d%/%m%/%Y% %H%:%M%:%S% %from(name)% <%from(mail)%> wrote:
> To: %to(name)% <%to(mail)%>[[
> Cc: %cc(name)% <%cc(mail)%>]]
> Subject: %subject%
> Message-ID: %Message-ID%
[[> Reply-to: %Reply-to%
]]>
```

**Comment by RealRaven2000** [comment]
My result (If I force plain text mode by holding SHIFT while clicking on forward). Not pretty but seems to do what you want it to:

![image](https://user-images.githubusercontent.com/6314674/103445835-d8116980-4c70-11eb-95da-95dbe5b9f1b6.png)

**Comment by RealRaven2000** [comment]
By the way - a more economic way to write the first line would be:

`%X:=sent%On %dateformat(d/m/Y H:M:S)% %from(name,bracketMail())% wrote:
`

**Comment by RealRaven2000** [comment]
There is some special code (with an explicit comment) for attaching inline when the `notifyComposeBodyReady` event is received; this determines whether the parameter `startup=true` is passed into `insertTemplate()` function. Ultimately, startup being true the old template will not be removed. The variable `isChangeTemplate` is only passed in as true if the user manually selects a new template from the composer toolbar; if the identity changes caused by another Add-on this is not triggered.

```
      // guard against forwarding my own message (body may have the smartTemplateInserted flag already)
if ( !root.getAttribute('smartTemplateInserted') 
        || gMsgCompose.type == msgComposeType.ForwardInline
        || flags.isThunderbirdTemplate || isChangeTemplate)  {
  let isStartup = isChangeTemplate ? false : !flags.isThunderbirdTemplate;
...
  flags.isChangeTemplate = isChangeTemplate;
  this.smartTemplate.insertTemplate(isStartup, flags, fileTemplateSource); // if a Tb template is opened, process without removal

}
```

I don't remember why and how. but I have added some code that now tells the insertTemplate function that the template is changed if it contains that attribute (`smartTemplateInserted`). 

```
if ((gMsgCompose.type == msgComposeType.ForwardInline || gMsgCompose.type == msgComposeType.ForwardAsAttachment)
    && root.getAttribute('smartTemplateInserted'))
  isChangeTemplate = true;
```

So here is a new version for testing:

[smartTemplate-fx-3.3pre33.zip](https://github.com/RealRaven2000/SmartTemplate4/files/5759465/smartTemplate-fx-3.3pre33.zip)

hope that forwarding change doesn't come and bite me later...

**Comment by RealRaven2000** [comment]
Implemented in 3.3 released for Thunderbird 78 on 03/01/2021

**Comment by macdeport** [comment]
Version 3.3 clearly improves duplicate template syndrome but still some mails are affected...

**Comment by RealRaven2000** [comment]
> Version 3.3 clearly improves duplicate template syndrome but still some mails are affected...

Good - if you can come up with a reproducible test case - the best solution is exporting your account / id settings (you can send me in an email for privacy). Does it only happen when forwarding?

This is how to export templates for one account as a json file:
![image](https://user-images.githubusercontent.com/6314674/103538531-43845280-4e8e-11eb-9763-405a75cbe7ff.png)

**Comment by macdeport** [comment]
I have a precise case of this issue 7 hours ago but not yet ??? Let's leave time to Time because reality is stubborn...

**Comment by RealRaven2000** [comment]
Reopening for now. I still believe it is "mostly fixed", but I am open to do more work on this, obviously, wherever it doesn't work. Duplicating the template is a nuisance!

**Comment by macdeport** [comment]
> This is how to export templates for one account as a json file:

Interesting this _Save account config_ button and the next one _Load account config_

**Comment by RealRaven2000** [comment]
> > This is how to export templates for one account as a json file:
> 
> Interesting this _Save account config_ button and the next one _Load account config_

yes that's basically it - saves all Identity related configuration but none of the Global / Advanced settings.

**Comment by RealRaven2000** [comment]
Already fixed in 2.13 / 3.3
